docs/sphinx/configurations.rst
..
    This file is part of BDC-STAC.
    Copyright (C) 2022 INPE.

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program. If not, see <https://www.gnu.org/licenses/gpl-3.0.html>.

.. _conf:

Service Configuration
=====================


.. data:: SQLALCHEMY_DATABASE_URI

   The database URI that should be used for the database connection. Defaults to ``'postgresql://postgres:postgres@localhost:5432/bdc'``.


.. data:: SQLALCHEMY_TRACK_MODIFICATIONS

    Enable (True) or disable (False) signals before and after changes are committed to the database. Defaults to ``False``.


.. data:: SQLALCHEMY_ECHO

   Enables or disable debug output of statements to ``stderr``. Defaults to ``False``.


.. data:: SQLALCHEMY_ENGINE_OPTIONS

    Set SQLAlchemy engine options for pooling and recycle.
    You may set the following environment variables to customize pooling:

    - ``SQLALCHEMY_ENGINE_POOL_SIZE``: The pool size. Defaults to ``5``.
    - ``SQLALCHEMY_ENGINE_MAX_OVERFLOW``: Max pool overflow. Defaults to ``10``.
    - ``SQLALCHEMY_ENGINE_POOL_CLASS``: The pool type for management. Defaults to ``10``.
    - ``SQLALCHEMY_ENGINE_POOL_RECYCLE``: Define the given number of seconds to recycle pool. Defaults to ``-1``, or no timeout.

.. data:: BDC_STAC_BASE_URL

    The base URI of the service. Defaults to ``'http://localhost:5000'``.

    This value is ignored if ``X-Stac-Url`` is present in ``request.headers``.


.. data:: BDC_STAC_FILE_ROOT

    The prefix for image ``assets``. Defaults to ``'http://localhost:5001'``.

    This value is ignored if ``X-Script-Name`` is present in ``request.headers``.


.. data:: BDC_STAC_MAX_LIMIT

    The limit of items returned in a query. Defaults to ``1000`` (an integer value).


.. data:: BDC_AUTH_CLIENT_ID

    Client ID generated by BDC-Auth. Defaults to ``None``, that means only public collections will be returned.



.. data:: BDC_AUTH_CLIENT_SECRET

    Client secret generated by BDC-Auth. Defaults to ``None``, that means only public collections will be returned.


.. data:: BDC_AUTH_ACCESS_TOKEN_URL

    Access token url used for retrieving user info in BDC-Auth. Defaults to ``None``, that means only public collections will be returned.


.. data:: BDC_STAC_USE_FOOTPRINT

    Flag to set if Item intersection should use ``Item.footprint``. Defaults to ``0``, which means to use ``Item.bbox``.



CHANGES.rst
..
    This file is part of BDC-STAC.
    Copyright (C) 2022 INPE.

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program. If not, see <https://www.gnu.org/licenses/gpl-3.0.html>.


=======
Changes
=======

Version 1.0.2 (2023-05-17)
--------------------------

- Review way to represent deprecated collections and versioning (`#88 <https://github.com/brazil-data-cube/bdc-stac/issues/88>`_)
- Add fixture for tests and remove warnings PEP8
- Upgrade Docker base image for python:3.11
- Upgrade bdc-catalog to 1.0.2 to improve test coverage related successor/predecessor
- Add option ``bdc:public`` to display collection availability


Version 1.0.1 (2022-12-06)
--------------------------

- Add integration with ``BDC-Catalog`` 1.0.1 which improves item retrieval using PostgreSQL combined indexes.
- Add support of extension `storage <https://github.com/stac-extensions/storage>`_ `#177 <https://github.com/brazil-data-cube/bdc-stac/issues/177>`_.
- Fix bug related ``platform`` and ``instruments`` for collections/items `#178 <https://github.com/brazil-data-cube/bdc-stac/issues/178>`_.
- Review Dockerfile install dir and build arguments.


Version 1.0.0 (2022-10-06)
--------------------------

- Change LICENSE to GPL v3
- Add integration with ``BDC-Catalog`` 1.0 and ``BDC-Auth-Client`` 0.4.2
- Mark repository as ``stable``


Version 0.9.0-15 (2022-09-27)
-----------------------------

- Fix dependency version for GeoAlchemy2


Version 1.0.0a1 (2022-09-12)
----------------------------

- Implementation of STAC v1.0-rc.1 with the following extensions:

   - `Electro-Optical (eo) <https://github.com/stac-extensions/eo>`_;
   - `SAR <https://github.com/stac-extensions/sar>`_;
   - `Item Assets <https://github.com/stac-extensions/item-assets>`_;
   - `Datacube <https://github.com/stac-extensions/datacube>`_;
   - `Processing <https://github.com/stac-extensions/processing>`_.

- Review the database queries to improve performance
- Improve documentation: Running and deploy `#163 <https://github.com/brazil-data-cube/bdc-stac/issues/163>`_.
- Review unittests


Version 0.9.0-14 (2021-04-01)
-----------------------------

- Improve pagination when using POST in /search (`#141 <https://github.com/brazil-data-cube/bdc-stac/pull/141>`_)
- remove the question mark if there are no args in the url (`#147 <https://github.com/brazil-data-cube/bdc-stac/pull/147>`_)

Version 0.9.0-13 (2021-01-28)
-----------------------------

- Add drone support (`#133 <https://github.com/brazil-data-cube/bdc-stac/issues/133>`_)
- Remove collection metadata from item (`#136 <https://github.com/brazil-data-cube/bdc-stac/issues/136>`_)
- Add missing CORS headers (`#138 <https://github.com/brazil-data-cube/bdc-stac/issues/138>`_)
- Pass auth parameters to asset url (`#130 <https://github.com/brazil-data-cube/bdc-stac/issues/130>`_)
- Fix bug when access collection item id directly (`#140 <https://github.com/brazil-data-cube/bdc-stac/issues/140>`_)

Version 0.9.0-12 (2021-01-14)
-----------------------------

- Add configuration for catalog description and id. (`#129 <https://github.com/brazil-data-cube/bdc-stac/issues/129>`_)

Version 0.9.0-11 (2021-01-05)
-----------------------------


- Display end_datetime in every item. (`#124 <https://github.com/brazil-data-cube/bdc-stac/issues/124>`_)
- Fix datetime filter. (`#127 <https://github.com/brazil-data-cube/bdc-stac/issues/127>`_)


Version 0.9.0-10 (2020-12-01)
-----------------------------

- Add bdc extension schema. (`#89 <https://github.com/brazil-data-cube/bdc-stac/issues/89>`_)
- Add route to serve jsonschemas. (`#90 <https://github.com/brazil-data-cube/bdc-stac/issues/90>`_)
- Add collection_type to collection. (`#120 <https://github.com/brazil-data-cube/bdc-stac/issues/120>`_)
- Fix duplicated item on pagination. (`#121 <https://github.com/brazil-data-cube/bdc-stac/issues/121>`_)


Version 0.9.0-9 (2020-11-23)
-----------------------------

- Use timeline table. (`#87 <https://github.com/brazil-data-cube/bdc-stac/issues/87>`_)
- Fix datetime parameter to work as specified in (`OGC API <http://docs.opengeospatial.org/is/17-069r3/17-069r3.html#_parameter_datetime>`_).
- Improve documentation on datetime parameter (`#114 <https://github.com/brazil-data-cube/bdc-stac/issues/114>`_).


Version 0.9.0-8 (2020-11-16)
----------------------------


- Fix item metadata (`#117 <https://github.com/brazil-data-cube/bdc-stac/pull/117>`_).


Version 0.9.0-7 (2020-11-13)
----------------------------


- Check if platform exists in collection (`#116 <https://github.com/brazil-data-cube/bdc-stac/pull/116>`_).


Version 0.9.0-6 (2020-11-12)
----------------------------


- Display all metadata in collection (`#115 <https://github.com/brazil-data-cube/bdc-stac/pull/115>`_).


Version 0.9.0-5 (2020-10-26)
----------------------------


- Fix collection with no associated grid (`#112 <https://github.com/brazil-data-cube/bdc-stac/pull/112>`_).


Version 0.9.0-4 (2020-10-05)
----------------------------


- Add instrument and platform to collections and items.


Version 0.9.0-3 (2020-09-18)
----------------------------


- Bug fix: Review some links in the service routes (`#110 <https://github.com/brazil-data-cube/bdc-stac/pull/110>`_)



Version 0.9.0-2 (2020-09-11)
----------------------------


- Bug fix: Check if parameter ``bbox`` is valid (an area object) (`#105 <https://github.com/brazil-data-cube/bdc-stac/issues/105>`_)


Version 0.9.0-1 (2020-09-09)

bdc_stac/config.py
#
# This file is part of BDC-STAC.
# Copyright (C) 2022 INPE.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/gpl-3.0.html>.
#
"""Config module."""

import os
from distutils.util import strtobool

SQLALCHEMY_DATABASE_URI = os.getenv("SQLALCHEMY_DATABASE_URI", "postgresql://postgres:postgres@localhost:5432/bdc")
"""The database URI that should be used for the database connection. 
Defaults to ``'postgresql://postgres:postgres@localhost:5432/bdc'``."""

SQLALCHEMY_TRACK_MODIFICATIONS = os.getenv("SQLALCHEMY_TRACK_MODIFICATIONS", False)
"""Enable (True) or disable (False) signals before and after changes are committed to the database. 
Defaults to ``False``."""

SQLALCHEMY_ECHO = os.getenv("SQLALCHEMY_ECHO", False)
"""Enables or disable debug output of statements to ``stderr``. Defaults to ``False``."""

SQLALCHEMY_ENGINE_OPTIONS = {
    "pool_size": int(os.getenv("SQLALCHEMY_ENGINE_POOL_SIZE", 5)),
    "max_overflow": int(os.getenv("SQLALCHEMY_ENGINE_MAX_OVERFLOW", 10)),
    "poolclass": os.getenv("SQLALCHEMY_ENGINE_POOL_CLASS"),
    "pool_recycle": int(os.getenv("SQLALCHEMY_ENGINE_POOL_RECYCLE", -1)),
}
"""Set SQLAlchemy engine options for pooling.
You may set the following environment variables to customize pooling:

- ``SQLALCHEMY_ENGINE_POOL_SIZE``: The pool size. Defaults to ``5``.
- ``SQLALCHEMY_ENGINE_MAX_OVERFLOW``: Max pool overflow. Defaults to ``10``.
- ``SQLALCHEMY_ENGINE_POOL_CLASS``: The pool type for management. Defaults to ``10``.
- ``SQLALCHEMY_ENGINE_POOL_RECYCLE``: Define the given number of seconds to recycle pool. Defaults to ``-1``, or no timeout.
"""

BDC_STAC_API_VERSION = os.getenv("BDC_STAC_API_VERSION", "1.0.0")
BDC_STAC_BASE_URL = os.getenv("BDC_STAC_BASE_URL", "http://localhost:5000")
BDC_STAC_FILE_ROOT = os.getenv("BDC_STAC_FILE_ROOT", "http://localhost:5001")
BDC_STAC_MAX_LIMIT = int(os.getenv("BDC_STAC_MAX_LIMIT", "1000"))
BDC_STAC_TITLE = os.getenv("BDC_STAC_TITLE", "Brazil Data Cube Catalog")
BDC_STAC_ID = os.getenv("BDC_STAC_ID", "bdc")
BDC_STAC_ASSETS_ARGS = os.getenv("BDC_STAC_ASSETS_ARGS", "access_token")
BDC_AUTH_CLIENT_SECRET = os.getenv("BDC_AUTH_CLIENT_SECRET", None)
BDC_AUTH_CLIENT_ID = os.getenv("BDC_AUTH_CLIENT_ID", None)
BDC_AUTH_ACCESS_TOKEN_URL = os.getenv("BDC_AUTH_ACCESS_TOKEN_URL", None)
BDC_STAC_USE_FOOTPRINT = strtobool(os.getenv("BDC_STAC_USE_FOOTPRINT", "0"))
"""Flag to set if Item intersection should use ``Item.footprint``.
Defaults to ``0``, which means to use ``Item.bbox``."""

STAC_GEO_MEDIA_TYPE = "application/geo+json"

STAC_EXTENSION_MAP = {
    "eo": f"https://stac-extensions.github.io/eo/v{BDC_STAC_API_VERSION}/schema.json",
    "storage": f"https://stac-extensions.github.io/eo/v{BDC_STAC_API_VERSION}/schema.json",
    "datacube": f"https://stac-extensions.github.io/datacube/v{BDC_STAC_API_VERSION}/schema.json",
    "processing": f"https://stac-extensions.github.io/processing/v{BDC_STAC_API_VERSION}/schema.json",
    "sar": f"https://stac-extensions.github.io/sar/v{BDC_STAC_API_VERSION}/schema.json",
    "item-assets": f"https://stac-extensions.github.io/item-assets/v{BDC_STAC_API_VERSION}/schema.json",
    "version": f"https://stac-extensions.github.io/version/v{BDC_STAC_API_VERSION}/schema.json",
}


def get_stac_extensions(*args):
    """Retrieve a list of STAC Extension URL."""
    out = [STAC_EXTENSION_MAP.get(ext) for ext in args]
    return out
